### ISSUE TITLE
Support non-C# applications

### ISSUE THREAD

**Issue by vadi2** [velopack/velopack#1]
Hi!

Will non-C# applications be supported as well? This is something that Squirrel did really well.

**Comment by caesay** [comment]
Out of curiosity, which language do you want to support?

The TLDR is, yes - maybe in the future, not at the moment.

The long answer is: 

It worked well with the original Squirrel because of it's decision to write new versions to `app-*` folders. Your app runs in `app-1.0.0` for example, and an update is downloaded to `app-2.0.0`. The shortcut for your apps run `Update.exe --processStart=yourExe.exe` and then the updater finds the latest version to launch. 

There are so many architectural problems with this that it's hard to count. Your app essentially can have two different versions running at the same time in different locations. So Clowd.Squirrel V3 - and now Velopack, choose to install your app into a `current` folder, which is carefully and atomicly replaced during updates. While we replace that dir, we can check for things like missing dependencies (say you took out a new dependency on vcredist for example). It solves firewall rules, shortcut/pinning issues, tray icons, and much more - since your app is now running from a constant location. 

However, this new setup means you can't "fire and forget" an update in quite the same way. The C# SDK is now a small library which interacts with the Update.exe binary as necessary to do this replacement smoothly and reliably. Ideally, to support non C# applications we would write similar SDK's to make sure that this works well out of the box and there is no fiddling for developers.

**Comment by vadi2** [comment]
C++. I did have to do the fiddling myself, but it wasn't too bad, works reliably and has been battle tested: https://github.com/Mudlet/Mudlet/blob/aef716b05c41be2dd1d1352edc3a95df40745f27/src/main.cpp#L624-L651

**Comment by caesay** [comment]
If I was to write a small C/C++ SDK would you be interested in using Velopack for mudlet on Windows and OSX? Or just Windows? And how would you want it shipped - as a header-only lib? or as a compiled .dll? 

On OSX, Velopack is capable of taking a build output and turning it into a signed and notarised app bundle and .pkg installer ready to go. If you are interested in having a chat, pop into our discord?

**Comment by vadi2** [comment]
Let's try it for both. We did have the OSX pipeline sorted out pretty well, but the [signing tool](https://github.com/mitchellh/gon) we use is no longer in active maintenance, which is a potential problem.

As for shipping - first preference would be that it is available on [vcpkg](https://vcpkg.io/en/), as I'm trying to use this C++ package manager for as many dependencies as possible. The second preference would be a header-only lib.

**Comment by caesay** [comment]
https://github.com/velopack/velopack/assets/1287295/5875e4cd-7341-4d20-890f-921d186c0c1f

Just a proof of concept for now. I decided to go the "single-header-file" route.

**Comment by vadi2** [comment]
Nice nice!

**Comment by caesay** [comment]
@vadi2 I have basically non-existent experience building cross-platform apps in C++, is this something you'd be willing to lend a hand with at all? If no time to directly contribute, would you be free for a 10 min call so I can ask some questions about the API design / unicode string handling?

**Comment by caesay** [comment]
This is now supported at https://github.com/velopack/velopack.fusion
